src/vsr/replica.zig

```diff
@@ -5783,6 +5786,27 @@ pub fn ReplicaType(
             ).?;
         }
 
+        /// Returns the highest op that this replica can safely prepare_ok.
+        ///
+        /// Sending prepare_ok for a op higher than `op_prepare_ok_max` causes a replica to falsely
+        /// contribute to the durability of a checkpoint:
+        /// * A replica could sync to a checkpoint that is not yet durable on a quorum of
+        ///   replicas. To avoid falsely contributing to checkpoint durability, syncing replicas
+        ///   must withhold some prepare_oks till they haven't synced all tables.
+        /// * Replicas can safely write two pipeline_prepare_queue_max worth of ops past
+        ///   the next checkpoint's trigger to their WAL (see `op_prepare_max`). To avoid falsely
+        ///   contributing to the next checkpoint's durability, they must not send prepare_oks for
+        ///   ops past the first pipeline_prepare_queue_max.
+        fn op_prepare_ok_max(self: *const Self) u64 {
```

Comment:
Hm, this comment is a bit confusing to me. It makes sense, given that I know the context, but I fear that, when I forget the context, the comment' won't help me much!

Let's maybe re-word it to state invariants positively and front-load the "why", as per tiger style?

Something along the lines of (needs further rewording and grammar work):

* sending `prepare_ok` for `op=N` signifies that a replica has a sufficiently fresh checkponit. Namely, if replcia's current checkpoint is `C`, it won't send prepare OK for ops larger than `C + checkpoint_op + compaction_interval + piepeline_queue`
* this is used to determine when it is safe to overwrite old prepares from the journal (see op_repair_min)
* maybe explain why constants work out the way they do. 

THen, the part about _syncing_ replcia could actually be moved into the body of the function, as it is _implementaiton_ of the contract

Reply:
Agreed, I wasn't very happy with this comment as well, it definitely needs a lot of context to understand properly... I haven't spent too much time trying to formalize this yet, it would be great to do that now. Thank you for the direction!

Reply:
Took a stab at it in https://github.com/tigerbeetle/tigerbeetle/pull/2236/commits/33c264b3adfc07c29ec0181c7d5afca590ed08cb! Do let me know if we can make it clearer or improve it further. 😄